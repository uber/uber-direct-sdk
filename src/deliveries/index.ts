import type {
  DeliveryQuoteReq,
  DeliveryReq,
  ListDeliveriesReq,
  PODReq,
  UpdateDeliveryReq,
} from './types';
import { fetchData, makeQueryString } from '../utils';

export class DeliveriesClient {
  accessToken: string;
  baseURL: string;

  constructor(accessToken: string) {
    this.accessToken = accessToken;

    const customerId = process.env.UBER_DIRECT_CUSTOMER_ID;
    if (!customerId) {
      throw new Error('Must include UBER_DIRECT_CUSTOMER_ID in environment variables');
    }
    this.baseURL = `https://api.uber.com/v1/customers/${customerId}`;
  }

  async createQuote(req: DeliveryQuoteReq) {
    return fetchData(`${this.baseURL}/delivery_quotes`, 'POST', this.accessToken, req);
  }

  async createDelivery(req: DeliveryReq) {
    return fetchData(`${this.baseURL}/deliveries`, 'POST', this.accessToken, req);
  }

  async getDelivery(deliveryId: string) {
    return fetchData(`${this.baseURL}/deliveries/${deliveryId}`, 'GET', this.accessToken);
  }

  async listDeliveries(options?: ListDeliveriesReq) {
    let url = `${this.baseURL}/deliveries`;
    if (options) {
      url += makeQueryString(options);
    }
    return fetchData(url, 'GET', this.accessToken);
  }

  async cancelDelivery(deliveryId: string) {
    return fetchData(`${this.baseURL}/deliveries/${deliveryId}/cancel`, 'POST', this.accessToken);
  }

  async updateDelivery(deliveryId: string, req: UpdateDeliveryReq) {
    return fetchData(`${this.baseURL}/deliveries/${deliveryId}`, 'POST', this.accessToken, req);
  }

  async proofOfDelivery(deliveryId: string, req: PODReq) {
    return fetchData(
      `${this.baseURL}/deliveries/${deliveryId}/proof-of-delivery`,
      'POST',
      this.accessToken,
      req
    );
  }
}

export const createDeliveriesClient = (accessToken: string) => {
  return new DeliveriesClient(accessToken);
};
