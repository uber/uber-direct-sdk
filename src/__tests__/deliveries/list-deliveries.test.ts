import { DeliveriesClient, createDeliveriesClient } from '../../deliveries';
import {
  listDeliveriesResp,
  listDeliveriesReq,
  accessToken,
  customerId,
  getExpectedHeaders,
} from '../../__fixtures__/index';

describe('listDeliveries', () => {
  let deliveriesClient: DeliveriesClient;
  beforeEach(() => {
    jest.resetModules();

    process.env.UBER_DIRECT_CUSTOMER_ID = customerId;
    deliveriesClient = createDeliveriesClient(accessToken);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should make a GET request with the no parameters', async () => {
    global.fetch = jest.fn().mockResolvedValueOnce({
      ok: true,
      json: jest.fn().mockResolvedValueOnce(listDeliveriesResp),
    });

    const resp = await deliveriesClient.listDeliveries();
    console.log(getExpectedHeaders(accessToken, "GET"));
    expect(global.fetch).toHaveBeenCalledWith(
      `${deliveriesClient.baseURL}/deliveries`,
      {
        method: 'GET',
        headers: getExpectedHeaders(accessToken, 'GET'),
        body: undefined,
      }
    );

    expect(resp).toEqual(listDeliveriesResp);
  });

  it('should make a GET request with the optional parameters', async () => {
    global.fetch = jest.fn().mockResolvedValueOnce({
      ok: true,
      json: jest.fn().mockResolvedValueOnce(listDeliveriesResp),
    });

    const resp = await deliveriesClient.listDeliveries(listDeliveriesReq);
    expect(global.fetch).toHaveBeenCalledWith(
      `${deliveriesClient.baseURL}/deliveries?filter=pending&limit=1&Offset=1`,
      {
        method: 'GET',
        headers: getExpectedHeaders(accessToken, 'GET'),
      }
    );

    expect(resp).toEqual(listDeliveriesResp);
  });

  it('should handle errors', async () => {
    global.fetch = jest.fn().mockRejectedValueOnce({
      response: {
        data: {
          code: 'service_unavailable',
          message: 'Service is currently unavailable.',
          kind: 'error',
        },
      },
    });

    await expect(deliveriesClient.listDeliveries()).rejects.toThrowErrorMatchingSnapshot();
  });
});
