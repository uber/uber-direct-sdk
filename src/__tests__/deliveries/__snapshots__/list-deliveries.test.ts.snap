// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listDeliveries should handle errors 1`] = `"[object Object]"`;
